src/puppetlabs/puppetdb/http/query.clj

```diff
@@ -263,34 +234,77 @@
   "Takes a ring request map and extracts the puppetdb query from the
   request. GET and POST are accepted, all other requests throw an
   exception"
-  [req]
+  [req param-spec]
   (convert-query-params
    (case (:request-method req)
      :get (get-req->query req)
      :post (post-req->query req)
-     (throw (IllegalArgumentException. "PuppetDB queries must be made via GET/POST")))))
+     (throw (IllegalArgumentException. "PuppetDB queries must be made via GET/POST")))
+    param-spec))
 
 (defn extract-query
   "Query handler that converts the incoming request (GET or POST)
   parameters/body to a pdb query map"
-  [handler]
+  [handler param-spec]
   (fn [{:keys [request-method body params puppetdb-query] :as req}]
-    (handler (assoc req :puppetdb-query (create-query-map req)))))
+    (handler (assoc req :puppetdb-query (create-query-map req param-spec)))))
+
+(defn validate-distinct-options!
+  "Validate the HTTP query params related to a `distinct_resources` query.  Return a
+  map containing the validated `distinct_resources` options, parsed to the correct
+  data types.  Throws `IllegalArgumentException` if any arguments are missing
+  or invalid."
+  [{:keys [distinct_start_time distinct_end_time distinct_resources] :as params}]
+  (let [distinct-params #{:distinct_resources :distinct_start_time
+                          :distinct_end_time}
+        params-present (filter distinct-params (kitchensink/keyset params))]
+    (condp = (set params-present)
+     #{}
+      params
+
+     distinct-params
+     (let [start (to-timestamp distinct_start_time)
+           end   (to-timestamp distinct_end_time)]
+       (when (some nil? [start end])
+         (throw (IllegalArgumentException.
+                 (str "query parameters 'distinct_start_time' and 'distinct_end_time' must be valid datetime strings: "
+                      distinct_start_time " " distinct_end_time))))
+       (merge params
+              {:distinct_resources (boolean distinct_resources)
+               :distinct_start_time start
+               :distinct_end_time   end}))
+
+     #{:distinct_start_time :distinct_end_time}
+     (throw
+       (IllegalArgumentException.
+         "'distinct_resources' query parameter must accompany parameters 'distinct_start_time' and 'distinct_end_time'"))
+     (throw
+       (IllegalArgumentException.
+         "'distinct_resources' query parameter requires accompanying parameters 'distinct_start_time' and 'distinct_end_time'")))))
+
+(defn warn-experimental
+  [entity {:keys [product-name]}]
```

Comment:
Wouldn't mind seeing an `:as globals` on this for documentation purposes. Or perhaps better, letting it take product-name directly and pulling it out of globals at the call site.